internal/scheduler/legacyscheduler.go

```diff
@@ -0,0 +1,619 @@
+			continue
+		}
+		for {
+			report, err := it.Next()
+			if err != nil {
+				return nil, nil, err
+			}
+			if report == nil { // We've processed all jobs for this queue.
+				delete(c.PriorityFactorByQueue, queue)
+				break
+			}
+
+			// Check overall per-round resource limits.
+			// Add the resource requests of this job to the total usage for this queue.
+			//
+			// TODO: Account for resource usage separately by priority.
+			jobTotalResourceRequests := common.TotalJobResourceRequest(report.Job)
+			roundResourcesCopy := roundResources.DeepCopy()
+			for resourceType, quantity := range jobTotalResourceRequests {
+				q := roundResourcesCopy.Resources[resourceType]
+				q.Add(quantity)
+				roundResourcesCopy.Resources[resourceType] = q
+			}
+			if exceeded, reason := c.exceedsResourceLimits(
+				ctx,
+				roundResourcesCopy,
+				c.SchedulingConfig.MaximalClusterFractionToSchedule,
+			); exceeded {
+				report.UnschedulableReason = reason + " (overall per scheduling round limit)"
+				if c.SchedulingReportsRepository != nil {
+					c.SchedulingReportsRepository.Add(queue, report)
+				}
+				continue
+			}
+
+			// Find a node for this job.
+			// We need to do this here in addition to above,
+			// since it may no longer be possible to assign this job to the node in the report
+			// if other jobs were scheduled onto that node in the interim.
+			//
+			// TODO: Only repeat this process if the node in the report no longer works.
+			podReport, err := c.selectNodeForPod(ctx, report.JobId, report.Job)
+			if err != nil {
+				return nil, nil, err
+			}
+			report.PodSchedulingReports = []*PodSchedulingReport{podReport}
+			if podReport.Node == nil {
+				report.UnschedulableReason = "pod does not fit on any node"
+				if c.SchedulingReportsRepository != nil {
+					c.SchedulingReportsRepository.Add(queue, report)
+				}
+				continue // Found no node for this job.
+			}
+
+			// Mark the job to be leased, and update resource accounting.
+			it.Lease(report)
+			jobsToLeaseByQueue[queue] = append(jobsToLeaseByQueue[queue], report.Job)
+			numJobsToLease++
+			roundResources = roundResourcesCopy
+			totalResourcesByQueue[queue] = report.TotalQueueResources
+			mostRecentSuccessfulJobSchedulingReportByQueue[queue] = report
+			if c.SchedulingReportsRepository != nil {
+				// Zero out the job spec to reduce memory usage.
+				report.Job = nil
+				c.SchedulingReportsRepository.Add(queue, report)
+			}
+			break
+		}
+	}
+
+	// Try to create leases.
+	jobs := make([]*api.Job, 0, numJobsToLease)
+	for queue, jobsToLease := range jobsToLeaseByQueue {
+
+		// TryLeaseJobs returns a list of jobs that were successfully leased.
+		// For example, jobs concurrently leased to another executor are skipped.
+		//
+		// TODO: Reports generated above will be incorrect if creating the lease fails.
+		successfullyLeasedJobs, err := c.JobRepository.TryLeaseJobs(c.ExecutorId, queue, jobsToLease)
+		if err != nil {
```

Comment:
I think that you can fail to lease jobs if another instance beat you to it.  In this case logging errors with stacktraces is a bit disingenuous.  I guess the old implementation did something similar as the  problem  is with the JobRepo interface, but we should probably fix in a subsequent pr.

Reply:
Yes, I agree